File: tests/test_mpi3rma_fortran/get_acc_sum_pscw.f90
      subroutine test_mpi3rma_get_acc_sum_pscw( base, array_size, DIM1, win, comm )

      use utilities, only: check_fompi_status

#if INCLUDE_TYPE==1
      implicit none
      
      include 'fompif.h'
#endif
#if INCLUDE_TYPE==2
      use fompi
      
      implicit none
#endif

      integer, intent(in) :: array_size
      integer, dimension(array_size), intent(inout) :: base
      integer, intent(in) :: DIM1
      integer, intent(in) :: win
      integer, intent(in) :: comm

      integer :: i
      integer :: rank
      integer :: commsize
      integer :: mpierr
      integer(kind=MPI_ADDRESS_KIND) :: start
  
      logical :: flag
      integer(kind=MPI_ADDRESS_KIND) :: create_flavor

      integer, dimension(:), allocatable :: recv_buf

      integer, dimension(3) :: ranges
      integer :: group
      integer :: newgroup

      character(50) :: filename

      write (filename,*) "get_acc_sum_pscw.f90"

      call MPI_Comm_rank( comm, rank, mpierr )
      call MPI_Comm_size( comm, commsize, mpierr )

      allocate( recv_buf(DIM1), stat=mpierr )

      if ( rank .eq. 0 ) then
        ranges(1) = 1
        ranges(2) = commsize-1
        ranges(3) = 1
      else
        ranges(1) = 0
        ranges(2) = 0
        ranges(3) = 1
      endif
      call MPI_Comm_group(comm, group, mpierr)
      call MPI_Group_range_incl(group, 1, ranges, newgroup, mpierr)
      call MPI_Group_free( group, mpierr )

      call foMPI_Win_get_attr( win, foMPI_WIN_CREATE_FLAVOR, create_flavor, flag, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 49, rank)
      if (flag .neqv. .true.) then
        write (*,*) "Didn't found foMPI_WIN_CREATE_FLAVOR keyval."
        STOP
      endif

      if ( create_flavor .eq. foMPI_WIN_FLAVOR_DYNAMIC ) then
        if ( rank .eq. 0 ) then
          call foMPI_Get_address( base(1), start, mpierr )
        endif
        call MPI_Bcast( start, 1, MPI_AINT, 0, comm, mpierr )
      else
        start = 0
      endif

      do i=1,DIM1
        base(i) = i-1
      enddo

      if( rank .eq. 0 ) then
        call foMPI_Win_post(newgroup, 0, win, mpierr)
        call check_fompi_status(mpierr, MPI_SUCCESS, filename, 70, rank)
    
        call foMPI_Win_wait(win, mpierr);
        call check_fompi_status(mpierr, MPI_SUCCESS, filename, 73, rank)
      else
        call foMPI_Win_start(newgroup, 0, win, mpierr)
        call check_fompi_status(mpierr, MPI_SUCCESS, filename, 76, rank)

        call foMPI_Get_accumulate( base(1), DIM1, MPI_INTEGER, recv_buf(1), DIM1, MPI_INTEGER, 0, start, DIM1, MPI_INTEGER, &
          foMPI_SUM, win, mpierr )
        call check_fompi_status(mpierr, MPI_SUCCESS, filename, 79, rank)
    
        call foMPI_Win_complete(win, mpierr)
        call check_fompi_status(mpierr, MPI_SUCCESS, filename, 82, rank)
      endif

      call MPI_Group_free( newgroup, mpierr )

      if ( rank .eq. 0 ) then
        do i=1,DIM1
          if( base(i) .ne. (commsize*(i-1)) ) then
            write (*,*) "Error on rank ", rank, " in ", trim(filename), &
              " line ", 89, ": element ", i, " is ", base(i), &
              " instead of ", commsize*(i-1), "."
            STOP
          endif
        enddo
      else
        if( recv_buf(1) .ne. 0) then
          write (*,*) "Error on rank ", rank, " in ", trim(filename), &
            " line ", 99, ": element ", 1, " is ", recv_buf(1), &
            " instead of ", 0, "."
          STOP
        endif
        do i=2,DIM1
          if( mod(recv_buf(i),i-1) .ne. 0) then
            write (*,*) "Error on rank ", rank, " in ", trim(filename), &
              " line ", 99, ": element ", i, " is ", mod(recv_buf(i),i-1), &
              " instead of ", 0, "."
            STOP
          endif
        enddo
      endif

      deallocate( recv_buf, stat=mpierr )

      call MPI_Barrier( comm, mpierr )

      if (rank .eq. 0) then
        write (*,*) "No Error"
      endif

      end subroutine test_mpi3rma_get_acc_sum_pscw

File: tests/test_foMPI-NA/get_acc_sum_pscw.c
#include <assert.h>

#include "fompi.h"
#include "test_mpi3rma.h"

void test_mpi3rma_get_acc_sum_pscw( int* base, int DIM, foMPI_Win win, MPI_Comm comm ) {

  int i;
  int rank;
  int commsize;
  int status;
  MPI_Aint start;

  int flag, create_flavor;

  int* recv_buf;
  int ranges[1][3];
  MPI_Group group;
  MPI_Group newgroup;

  MPI_Comm_rank( comm, &rank );
  MPI_Comm_size( comm, &commsize );

  recv_buf = malloc( DIM * sizeof(int) );

  if ( rank == 0 ) {
    ranges[0][0] = 1;
    ranges[0][1] = commsize-1;
    ranges[0][2] = 1;
  } else {
    ranges[0][0] = 0;
    ranges[0][1] = 0;
    ranges[0][2] = 1;
  }
  MPI_Comm_group(comm, &group);
  MPI_Group_range_incl(group, 1, ranges, &newgroup);
  MPI_Group_free( &group );

  status = foMPI_Win_get_attr( win, foMPI_WIN_CREATE_FLAVOR, &create_flavor, &flag );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
  assert( flag != 0 );

  if ( create_flavor == foMPI_WIN_FLAVOR_DYNAMIC ) {
    if ( rank == 0 ) {
      MPI_Get_address( &base[0], &start);
    }
    MPI_Bcast( &start, 1, MPI_AINT, 0, comm );
  } else {
    start = 0;
  }

  for( i=0 ; i<DIM ; i++ ) {
    base[i] = i;
  }

  if( rank == 0 ) {
    status = foMPI_Win_post(newgroup, 0, win);
    _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
    
    status = foMPI_Win_wait(win);
    _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
  } else {
    status = foMPI_Win_start(newgroup, 0, win);
    _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
    
    status = foMPI_Get_accumulate( &base[0], DIM, MPI_INT, &recv_buf[0], DIM, MPI_INT, 0, start, DIM, MPI_INT, foMPI_SUM, win );
    _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
    
    status = foMPI_Win_complete(win);
    _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
  }

  MPI_Group_free( &newgroup );

  if ( rank == 0 ) {
    for( i=0 ; i<DIM ; i++ ) {
      if( base[i] != (commsize*i) ) {
        fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, i, base[i], commsize*i );
        exit(EXIT_FAILURE);
      }
    }
  } else {
    if( recv_buf[0] != 0) {
      fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, 0, recv_buf[0], 0 );
      exit(EXIT_FAILURE);
    }
    for( i=1 ; i<DIM ; i++ ) {
      if( (recv_buf[i] % i) != 0) {
        fprintf(stderr, "Error on rank %i in %s line %i: element %i (%i) is not a multiple of %i.\n", rank, __FILE__, __LINE__, i, recv_buf[i], i );
        exit(EXIT_FAILURE);
      }
    }
  }

  free( recv_buf );

  MPI_Barrier( comm );

  if (rank == 0) {
    printf("No Error\n");
  }

  return;
}

